codeworld-api/src/CodeWorld/Driver.hs

```diff
@@ -2540,12 +2553,44 @@ simulationControls w
     | playbackSpeed w == 0 = [PlayButton, StepButton, SpeedSlider]
     | otherwise = [PauseButton, SpeedSlider]
 
+statefulDebugControls :: Wrapped [w] -> [Control [w]]
+statefulDebugControls w
+    | lastInteractionTime w > 5 = []
+    | playbackSpeed w == 0 && not (null (tail (state w))) = [PlayButton, StepButton, SpeedSlider, UndoButton]
+    | playbackSpeed w == 0 = [PlayButton, StepButton, SpeedSlider]
+    | otherwise = [PauseButton, SpeedSlider]
+
 simulationOf simInitial simStep simDraw =
-    runPauseable initial simStep simulationControls simDraw
+    runInspect simulationControls initial simStep (\_ r -> r) simDraw
   where
     initial =
         Wrapped {state = simInitial, playbackSpeed = 1, lastInteractionTime = 1000, isDragging = False}
 
+debugSimulationOf simInitial simStep simDraw =
+    runInspect statefulDebugControls initial step (\_ r -> r) draw
+  where
+    initial =
+        Wrapped {state = [simInitial], playbackSpeed = 1, lastInteractionTime = 1000, isDragging = False}
+    step dt (x:xs) = case x `seq` x' `seq` reallyUnsafePtrEquality# x x' of
+        0# -> x' : x : xs
+        _  -> x : xs
+      where x' = simStep dt x
+    draw (x:xs) = simDraw x
+
 trace msg x = unsafePerformIO $ do
```

Comment:
It doesn't make sense to put trace between the two debug entry points.  Put it back at the end of the file?